course_discovery/apps/course_metadata/management/commands/backfill_course_run_slugs_to_courses.py

```
@@ -0,0 +1,62 @@
+
+from course_discovery.apps.course_metadata.models import (
+    BackfillCourseRunSlugsConfig, CourseRun, CourseRunStatus, CourseUrlSlug
+)
+
+logger = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
+    """ Management command to add redirects from course run slugs to courses"""
+
+    help = 'Adds published course run slugs to courses if not already present'
+
+    def add_arguments(self, parser):
+        parser.add_argument('--all', action='store_true', help='Add redirects from all course run slugs')
+        parser.add_argument('-uuids', nargs="*", help='Add redirects from all course run slugs for specific courses')
+        parser.add_argument('--args-from-database', action='store_true',
+                            help=('Use arguments from the BackfillCourseRunSlugsConfig model instead of the '
+                                  'command line.')
+                            )
+
+    def handle(self, *args, **options):
+        # using mutually exclusive argument groups in management commands is only supported in Django 2.2
+        # so use XOR to check manually
+        if not bool(options['args_from_database']) ^ (bool(options['uuids']) ^ bool(options['all'])):
+            raise CommandError('Invalid arguments')
+        options_dict = options
+        if options_dict['args_from_database']:
+            options_dict = self.get_args_from_database()
+        self.add_redirects_from_course_runs(**options_dict)
+
+    def add_course_run_redirect_to_parent_course(self, course_run):
+        if course_run.status != CourseRunStatus.Published or course_run.draft:
+            return
+        try:
+            course_run.course.url_slug_history.update_or_create(
+                url_slug=course_run.slug, course=course_run.course, partner=course_run.course.partner
+            )
+        except IntegrityError:
+            existing_slug = CourseUrlSlug.objects.get(url_slug=course_run.slug, partner=course_run.course.partner)
```

Alternatively it might just be better to do a pre-check. At the very least with this approach, I'd want to know what the IntegrityError is (uniqueness/fk constraint) failure so we aren't just swallowing errors that exist and just assume it's from a slug belonging to a course already. But again, if this is our normal and expected path for some of these courses to take, I'd also rather not see stacktraces for non actionable errors.